File: app/src/main/java/org/mifos/mobile/ui/fragments/NotificationFragment.java
package org.mifos.mobile.ui.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler;

import org.mifos.mobile.R;
import org.mifos.mobile.models.notification.MifosNotification;
import org.mifos.mobile.presenters.NotificationPresenter;
import org.mifos.mobile.ui.activities.base.BaseActivity;
import org.mifos.mobile.ui.adapters.NotificationAdapter;
import org.mifos.mobile.ui.fragments.base.BaseFragment;
import org.mifos.mobile.ui.views.NotificationView;
import org.mifos.mobile.utils.DividerItemDecoration;
import org.mifos.mobile.utils.Network;

import java.util.List;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by dilpreet on 13/9/17.
 */

public class NotificationFragment extends BaseFragment implements NotificationView,
        SwipeRefreshLayout.OnRefreshListener {


    @BindView(R.id.rv_notifications)
    RecyclerView rvNotification;

    @BindView(R.id.swipe_notification_container)
    SwipeRefreshLayout swipeNotificationContainer;

    @BindView(R.id.layout_error)
    View layoutError;

    @Inject
    NotificationPresenter presenter;

    @Inject
    NotificationAdapter adapter;

    private View rootView;
    private SweetUIErrorHandler sweetUIErrorHandler;


    public static NotificationFragment newInstance() {
        NotificationFragment fragment = new NotificationFragment();
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setToolbarTitle(getString(R.string.notification));
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);

    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_notification, container, false);
        ButterKnife.bind(this, rootView);

        sweetUIErrorHandler = new SweetUIErrorHandler(getActivity(), rootView);

        LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity());
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        rvNotification.setLayoutManager(layoutManager);
        rvNotification.addItemDecoration(new DividerItemDecoration(getActivity(),
                layoutManager.getOrientation()));
        rvNotification.setAdapter(adapter);

        swipeNotificationContainer.setColorSchemeResources(R.color.blue_light,
                R.color.green_light, R.color.orange_light, R.color.red_light);
        swipeNotificationContainer.setOnRefreshListener(this);

        presenter.attachView(this);
        presenter.loadNotifications();

        return rootView;
    }

    @Override
    public void showNotifications(List<MifosNotification> notifications) {
        if (notifications.size() != 0) {
            adapter.setNotificationList(notifications);
        } else {
            sweetUIErrorHandler.showSweetEmptyUI(getString(R.string.notification),
                    R.drawable.ic_notifications, rvNotification, layoutError);
        }
    }

    @Override
    public void showError(String msg) {
        if (!Network.isConnected(getActivity())) {
            sweetUIErrorHandler.showSweetNoInternetUI(rvNotification, layoutError);
        } else {
            sweetUIErrorHandler.showSweetErrorUI(msg,
                    rvNotification, layoutError);
            Toast.makeText(getActivity(), msg, Toast.LENGTH_SHORT).show();
        }
    }

    @OnClick(R.id.btn_try_again)
    public void retryClicked() {
        if (Network.isConnected(getContext())) {
            sweetUIErrorHandler.hideSweetErrorLayoutUI(rvNotification, layoutError);
            presenter.loadNotifications();
        } else {
            Toast.makeText(getContext(), getString(R.string.internet_not_connected),
                    Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void showProgress() {
        swipeNotificationContainer.setRefreshing(true);
    }

    @Override
    public void hideProgress() {
        swipeNotificationContainer.setRefreshing(false);
    }

    @Override
    public void onRefresh() {
        sweetUIErrorHandler.hideSweetErrorLayoutUI(rvNotification, layoutError);
        presenter.loadNotifications();

    }
}

File: app/src/main/java/org/mifos/mobile/ui/fragments/NotificationFragment.kt
package org.mifos.mobile.ui.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler
import dagger.hilt.android.AndroidEntryPoint
import org.mifos.mobile.BuildConfig
import org.mifos.mobile.utils.NotificationUiState
import org.mifos.mobile.R
import org.mifos.mobile.databinding.FragmentNotificationBinding
import org.mifos.mobile.models.notification.MifosNotification
import org.mifos.mobile.ui.adapters.NotificationAdapter
import org.mifos.mobile.ui.fragments.base.BaseFragment
import org.mifos.mobile.utils.DividerItemDecoration
import org.mifos.mobile.utils.Network
import org.mifos.mobile.viewModels.NotificationViewModel
import javax.inject.Inject

/**
 * Created by dilpreet on 13/9/17.
 */
@AndroidEntryPoint
class NotificationFragment : BaseFragment(), OnRefreshListener {
    private var _binding: FragmentNotificationBinding? = null
    private val binding get() = _binding!!
    private lateinit var viewModel : NotificationViewModel

    @JvmField
    @Inject
    var adapter: NotificationAdapter? = null
    private var sweetUIErrorHandler: SweetUIErrorHandler? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setToolbarTitle(getString(R.string.notification))
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentNotificationBinding.inflate(inflater, container, false)
        val rootView = binding.root
        viewModel = ViewModelProvider(this)[NotificationViewModel::class.java]
        sweetUIErrorHandler = SweetUIErrorHandler(activity, rootView)
        val layoutManager = LinearLayoutManager(activity)
        layoutManager.orientation = LinearLayoutManager.VERTICAL
        binding.rvNotifications.layoutManager = layoutManager
        binding.rvNotifications.addItemDecoration(
            DividerItemDecoration(
                activity,
                layoutManager.orientation,
            ),
        )
        binding.rvNotifications.adapter = adapter
        binding.swipeNotificationContainer.setColorSchemeResources(
            R.color.blue_light,
            R.color.green_light,
            R.color.orange_light,
            R.color.red_light,
        )
        binding.swipeNotificationContainer.setOnRefreshListener(this)
        loadNotifications()
        return rootView
    }

    private fun showNotifications(notifications: List<MifosNotification?>?) {
        if (BuildConfig.DEBUG && notifications == null) {
            error("Assertion failed")
        }
        if (notifications?.isNotEmpty() == true) {
            adapter?.setNotificationList(notifications as List<MifosNotification?>?)
        } else {
            sweetUIErrorHandler?.showSweetEmptyUI(
                getString(R.string.notification),
                R.drawable.ic_notifications,
                binding.rvNotifications,
                binding.layoutError.root,
            )
        }
    }

    fun showError(msg: String?) {
        if (!Network.isConnected(activity)) {
            sweetUIErrorHandler?.showSweetNoInternetUI(
                binding.rvNotifications,
                binding.layoutError.root,
            )
        } else {
            sweetUIErrorHandler?.showSweetErrorUI(
                msg,
                binding.rvNotifications,
                binding.layoutError.root,
            )
            Toast.makeText(activity, msg, Toast.LENGTH_SHORT).show()
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel.notificationUiState.observe(viewLifecycleOwner) { state ->
            when(state) {
                NotificationUiState.Loading -> showProgress()
                NotificationUiState.Error -> {
                    hideProgress()
                    showError(context?.getString(R.string.notification))
                }
                is NotificationUiState.LoadNotificationsSuccessful -> {
                    hideProgress()
                    showNotifications(state.notifications)
                }
            }
        }

        binding.layoutError.btnTryAgain.setOnClickListener {
            retryClicked()
        }
    }

    private fun loadNotifications() {
        viewModel.loadNotifications()
    }

    private fun retryClicked() {
        if (Network.isConnected(context)) {
            sweetUIErrorHandler?.hideSweetErrorLayoutUI(
                binding.rvNotifications,
                binding.layoutError.root,
            )
            loadNotifications()
        } else {
            Toast.makeText(
                context,
                getString(R.string.internet_not_connected),
                Toast.LENGTH_SHORT,
            ).show()
        }
    }

    fun showProgress() {
        binding.swipeNotificationContainer.isRefreshing = true
    }

    fun hideProgress() {
        binding.swipeNotificationContainer.isRefreshing = false
    }

    override fun onRefresh() {
        sweetUIErrorHandler?.hideSweetErrorLayoutUI(
            binding.rvNotifications,
            binding.layoutError.root,
        )
        loadNotifications()
    }

    override fun onDestroyView() {
        _binding = null
        super.onDestroyView()
    }

    companion object {
        @JvmStatic
        fun newInstance(): NotificationFragment {
            return NotificationFragment()
        }
    }
}
